bootstrap/contracts/Deployer.sol

```
@@ -129,125 +166,171 @@ contract Deployer {
+    function addApp(bytes32 storageHash, bytes32 storageReceipt, uint8 clusterSize, bytes32[] pinToNodes)
         external
     {
-        enqueuedCodes.push(Code(storageHash, storageReceipt, clusterSize, msg.sender));
+        require(clusterSize >= pinToNodes.length,
+            "number of pinned nodes should be less or equal to the desired clusterSize");
 
-        if (!matchWork()) {
-            emit CodeEnqueued(storageHash);
-        }
+        App memory app = App(storageHash, storageReceipt, clusterSize, msg.sender, pinToNodes);
+
+        tryToDeployApp(app);
     }
 
-    /** @dev Checks if there is enough free Solvers for undeployed Code
-     * emits ClusterFormed event if so
-     */
-    function matchWork()
+    function tryToDeployApp(App memory app)
         internal
-        returns (bool)
+    returns(bool)
     {
-        uint idx = 0;
-        Code memory code;
+        Node[] memory solvers = new Node[](app.clusterSize);
+        uint8 solversLength = 0;
+
+        bool skip = false;
 
-        // TODO: better control enqueuedCodes.length so we don't exceed gasLimit
+        uint8 i = 0;
+        uint j = 0;
+        Node memory node;
 
-        // looking for a code that can be deployed given current number of readyNodes
-        for (; idx < enqueuedCodes.length; idx++) {
-            code = enqueuedCodes[idx];
-            if (readyNodes.length >= code.clusterSize) {
-                // suitable code found, stop on current idx
-                break;
+        for(; i < app.pinToNodes.length; i++) {
+            skip = false;
+            // It should be better then making a custom data structure due to small expected size of pinToNodes
+            for(; j < i && !skip; j++) {
+                skip = app.pinToNodes[i] == app.pinToNodes[j];
+            }
+            if(skip) {
+                continue;
+            }
+            node = nodes[app.pinToNodes[i]];
+            if(node.owner == msg.sender) {
```

can we throw an error, if not equals?